### PR TITLE
feat: add withdraw cap

### PR DESCRIPTION
**Pull request by MeLlamoPablo**
This PR implements the withdraw cap.

https://user-images.githubusercontent.com/11708035/220413032-1bf7828e-9d89-42ea-8d48-2936e7577784.mp4

For the example above, a mock withdraw cap of $100 per 24 hours is used. The PR uses a mock implementation of `WithdrawHook` deployed [here](https://goerli.etherscan.io/address/0x6b50c9bdac1f4b075cead4d1dbea80f92daba9d6#code). For testing, the `fakeAdvanceTime`, `fakeWithdraw` and `fakeReset` methods can be used.

The withdraw cap works similarly to the deposit cap, and reuses the `getBalanceLimitInfo` utility to compute whether the cap is `not-exceeded`, `already-exceeded` or `exceeded-after-transfer`.

# FE PR Checklist

- [x] Self-reviewed each line of code, preempted questions about logic or decisions
- [x] Noted any intentional deviations from design
- [x] Tried hard to break it and failed, think like a malicious user trying to find edge cases to break the app
- [x] Tested on supported device sizes
    - [x] iPhone 12 Pro
    - [x] iPad Air
    - [x] 1920x1080 (how to add custom size: [https://umaar.com/dev-tips/51-add-new-device/](https://umaar.com/dev-tips/51-add-new-device/))
- [x] Tested all widths between 1920px wide to 390px wide and confirmed no weird breakages at widths in between
- [x] (if related to balances) Tested with empty wallet, modest wallet, whale wallet
- [ ] (if browser specific logic) Tested on Brave, Firefox, Safari
- [x] (if styling) Tested Light/Dark mode

### CODE REVIEW

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts [hunk 1]
```diff
@@ -126,4 +151,47 @@ export class WithdrawStore {
...
+      formatUnits: this.root.preCTTokenStore.formatUnits.bind(this.root.preCTTokenStore),
+    })
+
+    if (limitInfo.status === 'already-exceeded' || limitInfo.status === 'exceeded-after-transfer') {
+      const nextGlobalPeriodReset = addDuration(lastGlobalPeriodReset, globalPeriodLength)
+      const timeToReset = differenceInMilliseconds(nextGlobalPeriodReset, nowInMs) as DurationInMs
+
+      return {
+        ...limitInfo,
+        resetsIn: periodAlreadyReset ? undefined : timeToReset,
```
Comment (chrisling-dev) [inline apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts:189]: if periodAlreadyReset is returned above with status = 'not-exceeded', do we still need this check here?

```suggestion
        resetsIn: timeToReset,
```

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawWarning.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      const cap = compactNumber(+withdrawLimit.capUnits, { showUsdSign: true })
+
+      return (
+        <Alert
+          message={
+            <>
+              <p>Global withdraw limit reached ({cap})</p>
+              {withdrawLimit.resetsIn !== undefined && (
+                <p>Resets in {formatDuration(withdrawLimit.resetsIn)}</p>
+              )}
```
Comment (chrisling-dev) [inline apps/frontend/core-dapp/src/features/withdraw/WithdrawWarning.tsx:24]: would resetsIn ever be undefined here for both 'already-exceeded' and 'exceeded-after-transfer' if combined with the solutions I provided earlier?

**Review by chrisling-dev — CHANGES_REQUESTED** [review]

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts [hunk 3]
```diff
@@ -126,4 +151,47 @@ export class WithdrawStore {
...
+    const periodAlreadyReset =
+      differenceInMilliseconds(nowInMs, lastGlobalPeriodReset) > globalPeriodLength
+
+    const limitInfo = getBalanceLimitInfo({
+      additionalAmount: this.withdrawalAmountBN,
+      cap: globalWithdrawLimitPerPeriod,
+      // If the reset window has passed, disregard the value of globalAmountWithdrawnThisPeriod.
+      // The amount withdrawn is effectively zero.
+      // When someone withdraws, globalAmountWithdrawnThisPeriod will update and thus the withdraw limit will be recomputed
+      currentAmount: periodAlreadyReset ? BigNumber.from(0) : globalAmountWithdrawnThisPeriod,
```
[the remark this answers is not in the document]
Comment (MeLlamoPablo) [reply apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts:177]: Even if `periodAlreadyReset = true`, it is possible that the user would want to withdraw more than `globalWithdrawLimitPerPeriod` (in which case the status would be `exceeded-after-transfer`)

If we returned early `not-exceeded` here we wouldn't prevent the user from making a deposit bigger than `globalWithdrawLimitPerPeriod`

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts (hunk 1 above)
Comment (MeLlamoPablo) [reply apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts:187]: As explained above, we can't return `not-exceeded` if `periodAlreadyReset`

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawWarning.tsx (hunk 2 above)
Comment (MeLlamoPablo) [reply apps/frontend/core-dapp/src/features/withdraw/WithdrawWarning.tsx:24]: The case where `resetsIn` can be undefined is: (considering a period of 24h and a limit of `$100`)

* Last reset: 3 days ago (therefore we are in a new period)
* I'm trying to withdraw `$200`

Therefore the cap info is:

* `status = exceeded-after-deposit`
* `resetsIn = undefined`

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts [hunk 4]
```diff
@@ -126,4 +151,47 @@ export class WithdrawStore {
...
+      formatUnits: this.root.preCTTokenStore.formatUnits.bind(this.root.preCTTokenStore),
```
[the remark this answers is not in the document]
Comment (MeLlamoPablo) [reply apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts:178]: It doesn't work without `.bind` for me

Screenshot 2023-02-22 at 13 47 36

[the remark this answers is not in the document]
Comment (chrisling-dev) [reply apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts:178]: oo okie thanks for checking!

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisling-dev) [reply apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts:177]: ahh okay I based my logic off an older version of the code where amount doesn't matter if you're not within the withdrawal window, this makes sense then

the older version logic:
Screenshot 2023-02-23 at 3 24 18 PM

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawWarning.tsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+  switch (withdrawLimit.status) {
+    case 'already-exceeded': {
+      const cap = compactNumber(+withdrawLimit.capUnits, { showUsdSign: true })
+
+      return (
+        <Alert
+          message={
+            <>
+              <p>Global withdraw limit reached ({cap})</p>
```
Comment (chrisling-dev) [inline apps/frontend/core-dapp/src/features/withdraw/WithdrawWarning.tsx:21]:
```suggestion
              <p>Global withdrawal limit reached ({cap})</p>
              <br />
```

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawWarning.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    case 'exceeded-after-transfer': {
+      const formattedRemainingAmount = formatUsd(+withdrawLimit.remainingUnits)
+
+      return (
+        <Alert
+          message={
+            <Flex flexDirection="column" gap={8} alignItems="start">
+              <p>
+                Withdraw limit exceeded. <br />
```
Comment (chrisling-dev) [inline apps/frontend/core-dapp/src/features/withdraw/WithdrawWarning.tsx:38]:
```suggestion
                Withdrawal limit exceeded. <br />
```

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawWarning.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                <InlineTextButton
+                  onClick={(): void => {
+                    setWithdrawalAmount(withdrawLimit.remainingUnits)
+                  }}
+                  title={`Withdraw ${formattedRemainingAmount} instead`}
+                >
+                  {formattedRemainingAmount}
+                </InlineTextButton>{' '}
+                remaining.
+              </p>
```
Comment (chrisling-dev) [inline apps/frontend/core-dapp/src/features/withdraw/WithdrawWarning.tsx:49]:
```suggestion
                (
                <InlineTextButton
                  onClick={(): void => {
                    setWithdrawalAmount(withdrawLimit.remainingUnits)
                  }}
                  title={`Withdraw ${formattedRemainingAmount} instead`}
                >
                  {formattedRemainingAmount}
                </InlineTextButton>{' '}
                remaining)
              </p>
```

**Review by chrisling-dev — APPROVED** [review]
LGTM after matching copy to design, well done!

File: apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (MeLlamoPablo) [reply apps/frontend/core-dapp/src/features/withdraw/WithdrawStore.ts:177]: Oh yeah, I saw that the audit placed a high severity issue on this, so I just assumed that this code would be changed

**Review by chrisling-dev — APPROVED** [review]
✅

**Pull request merged**